src/qs_dispersion_d4.F

```diff
@@ -0,0 +1,170 @@
+      CALL get_cell(cell=cell, periodic=periodic)
+      lperiod(1) = periodic(1) == 1
+      lperiod(2) = periodic(2) == 1
+      lperiod(3) = periodic(3) == 1
+
+      !prepare for the call to the dispersion function
+      CALL new(mol, el_num, xyz, lattice=cell%hmat, periodic=lperiod)
+      CALL new_d4_model(disp, mol)
+      CALL get_rational_damping(dispersion_env%ref_functional, param, s9=dispersion_env%s9)
+
+      ! Coordination number cutoff
+      cutoff%cn = dispersion_env%rc_cn
+      ! Two-body interaction cutoff
+      cutoff%disp2 = dispersion_env%rc_d4*2._dp
+      ! Three-body interaction cutoff
+      cutoff%disp3 = dispersion_env%rc_disp*2._dp
+
+      !> Wrapper to handle the evaluation of dispersion energy and derivatives
+      IF (PRESENT(force)) THEN
+         IF (PRESENT(atom_of_kind)) THEN
+            ALLOCATE (gradient(3, mol%nat))
+            CALL get_dispersion(mol, disp, param, cutoff, energy, gradient, stress)
+            IF (PRESENT(virial)) THEN
+               virial = stress
+            END IF
+            DO iatom = 1, natom
+               ikind = el_num(iatom)
+               ind_atom = atom_of_kind(iatom)
+               force(ikind)%dispersion(:, ind_atom) = force(ikind)%dispersion(:, ind_atom) + gradient(:, iatom)
+            END DO
+            DEALLOCATE (gradient)
+         ELSE
+            CPABORT("missing atom_of_kind")
+         END IF
+      ELSE
+         IF (PRESENT(virial)) THEN
+            CPABORT("missing force for virial term")
+         END IF
+         CALL get_dispersion(mol, disp, param, cutoff, energy)
+      END IF
+
+      DEALLOCATE (el_num, xyz)
+
+      CALL timestop(handle)
+
+   END SUBROUTINE calculate_dispersion_d4_pairpot
+
+#else
+
+   ! **************************************************************************************************
+!> \brief ...
+!> \param dispersion_env ...
+!> \param particle_set ...
+!> \param cell ...
+!> \param energy ...
+!> \param force ...
+!> \param atom_of_kind ...
+!> \param virial ...
+! **************************************************************************************************
+   SUBROUTINE calculate_dispersion_d4_pairpot(dispersion_env, particle_set, cell, energy, force, atom_of_kind, virial)
+
+      TYPE(qs_dispersion_type), INTENT(IN), POINTER      :: dispersion_env
+      TYPE(particle_type), DIMENSION(:), INTENT(IN), &
+         POINTER                                         :: particle_set
+      TYPE(cell_type), INTENT(IN), POINTER               :: cell
+      REAL(KIND=dp), INTENT(OUT)                         :: energy
+      TYPE(qs_force_type), DIMENSION(:), INTENT(INOUT), &
+         OPTIONAL, POINTER                               :: force
+      INTEGER, DIMENSION(:), INTENT(IN), OPTIONAL        :: atom_of_kind
+      REAL(KIND=dp), DIMENSION(3, 3), INTENT(OUT), &
+         OPTIONAL                                        :: virial
+
+      CPABORT("Build without DFTD4")
```

Comment:
Add `MARK_USED(variable)` for all dummy variables here to make the compiler(s) happy if DFTD4 is not in use.